src/box/box.cc

```diff
@@ -3555,23 +3589,47 @@ bootstrap(const struct tt_uuid *instance_uuid,
 	 * with connecting to 'replication_connect_quorum' masters.
 	 * If this also fails, throw an error.
 	 */
-	box_restart_replication();
-
-	struct replica *master = replicaset_find_join_master();
-	assert(master == NULL || master->applier != NULL);
-
-	if (master != NULL && !tt_uuid_is_equal(&master->uuid, &INSTANCE_UUID)) {
-		bootstrap_from_master(master);
-		/* Check replica set UUID */
-		if (!tt_uuid_is_nil(replicaset_uuid) &&
-		    !tt_uuid_is_equal(replicaset_uuid, &REPLICASET_UUID)) {
-			tnt_raise(ClientError, ER_REPLICASET_UUID_MISMATCH,
-				  tt_uuid_str(replicaset_uuid),
-				  tt_uuid_str(&REPLICASET_UUID));
+
+	struct replica *master;
+
+	/*
+	 * Rebootstrap
+	 *
+	 * 1) Try to connect to all nodes in the replicaset during
+	 *    the waiting period in order to update their ballot.
+	 *
+	 * 2) After updated the ballot of all nodes, it are looking
+	 *    for a new master.
+	 *
+	 * 3) Wait until bootstrap_from_master succeeds.
+	 */
+	while (true) {
+		box_restart_replication();
+		master = replicaset_find_join_master();
+
+		if (master != NULL &&
+		    !tt_uuid_is_equal(&master->uuid, &INSTANCE_UUID)) {
+			if (bootstrap_from_master_begin(master) == 0) {
+				bootstrap_from_master_end(master);
+				/* Check replica set UUID */
+				if (!tt_uuid_is_nil(replicaset_uuid) &&
+				    !tt_uuid_is_equal(replicaset_uuid,
+						      &REPLICASET_UUID)) {
+					tnt_raise(ClientError,
+						  ER_REPLICASET_UUID_MISMATCH,
+						  tt_uuid_str(replicaset_uuid),
+						  tt_uuid_str(&REPLICASET_UUID)
+						 );
+				}
+				break;
+			}
+		} else {
+			bootstrap_master(replicaset_uuid);
+			*is_bootstrap_leader = true;
+			break;
 		}
-	} else {
-		bootstrap_master(replicaset_uuid);
-		*is_bootstrap_leader = true;
+
+		fiber_sleep(0.1);
 	}
```

Comment:
Please,
1 - Reduce indentation level;
2 - Add some logging that a first join attempt has failed, and that you are retrying every N secs;
3 - It doesn't look a good idea to keep the retrying infinitely, better make it last not longer than `replication_connect_timeout`;

Reply:
We are still retrying infinitely despite I suggested to limit the attempts by `replication_connect_timeout`. But ok, I am fine with the current way too while nobody complains.